libraries/SITL/SIM_GPS.cpp

```diff
@@ -998,198 +996,447 @@ uint32_t GPS_NOVA::CalculateBlockCRC32(uint32_t length, uint8_t *buffer, uint32_
-    uint8_t buf[payload_sz] = {};
-    uint8_t offset = 0;
-    memcpy(&buf[offset], &pos_time, sizeof(pos_time));
-    offset += sizeof(pos_time);
-    memcpy(&buf[offset], &pos, sizeof(pos));
-    offset += sizeof(pos);
-    memcpy(&buf[offset], &vel, sizeof(vel));
-    offset += sizeof(vel);
-    memcpy(&buf[offset], &dop, sizeof(dop));
-    offset += sizeof(dop);
-    memcpy(&buf[offset], &pos_sigma, sizeof(pos_sigma));
-    offset += sizeof(pos_sigma);
     assert(offset == payload_sz);
-    send_gsof(buf, sizeof(buf));
+
+    // Don't send empy frames when all channels are disabled;
+    if (payload_sz > 0) {
+        send_gsof(buf, payload_sz);
+    }
+
+}
+
+bool DCOL_Parser::dcol_parse(const char data_in) {
+    bool ret = false;
+    switch(parse_state) {
+    case Parse_State::WAITING_ON_STX:
+        if (data_in == STX) {
+            reset();
+            parse_state = Parse_State::WAITING_ON_STATUS;
+        }
+        break;
+    case Parse_State::WAITING_ON_STATUS:
+        if (data_in != (uint8_t)Status::OK) {
+            // Invalid, status should always be OK.
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+        } else {
+            status = static_cast<Status>(data_in);
+            parse_state = Parse_State::WAITING_ON_PACKET_TYPE;
+        }
+        break;
+    case Parse_State::WAITING_ON_PACKET_TYPE:
+        if (data_in == (uint8_t)Packet_Type::COMMAND_APPFILE) {
+            packet_type = Packet_Type::COMMAND_APPFILE;
+        } else {
+            // Unsupported command in this simulator.
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+        }
+        parse_state = Parse_State::WAITING_ON_LENGTH;
+        break;
+    case Parse_State::WAITING_ON_LENGTH:
+        expected_payload_length = data_in;
+        parse_state = Parse_State::WAITING_ON_PACKET_DATA;
+        break;
+    case Parse_State::WAITING_ON_PACKET_DATA:
+        payload[cur_payload_idx] = data_in;
+        if (++cur_payload_idx == expected_payload_length) {
+            parse_state = Parse_State::WAITING_ON_CSUM;
+        }
+        break;
+    case Parse_State::WAITING_ON_CSUM:
+        expected_csum = data_in;
+        parse_state = Parse_State::WAITING_ON_ETX;
+        break;
+    case Parse_State::WAITING_ON_ETX:
+        if (data_in != ETX) {
+            reset();
+        }
+        if (!valid_csum()) {
+            // GSOF driver sent a packet with invalid CSUM.
+            // In real GSOF driver, the packet is simply ignored with no reply.
+            // In the SIM, treat this as a coding error.
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+        } else {
+            ret = parse_payload();
+        }
+        reset();
+        break;
+    }
+
+    return ret;
+}
+
+uint32_t DCOL_Parser::RateToPeriodMs(const Output_Rate rate) {
+    uint32_t min_period_ms = 0;
+    switch(rate) {
```

Comment:
nitpick: add a space between "space" and the "("

Reply:
Done!